daemontools-x/runpod.1
.\" vim: tw=75
.TH runpod 1
.SH NAME
runpod \- Run indimail podman / docker image using podman / docker

.SH SYNOPSIS
\fBrunpod\fR [\fIOPTIONS\fR] \fIprog\fR \fIarg\fR \fIarg\fR ...

.SH DESCRIPTION
\fBrunpod\fR is a frontend for \fBpodman\fR(1) / \fBdocker\fR(1) to help
execute indimail containers. indimail containers can be downloaded from

.nf
https://hub.docker.com/u/cprogrammer
or
https://github.com/orgs/indimail/packages
.fi

\fIprog\fR is usually the binary to be executed on startup. \fIprog\fR also
can have a special meaning. if \fIprog\fR is indimail, indimail-mta,
svscan, webmail, runpod will let \fBdocker-entrypoint\fR(8) execute
\fBsvscan\fR(8) in the container. This will start an services configured in
the \fB/service\fR directory of the container image. If you don't want
this, you can set \fBname\fR to execute something else instead of
\fBsvscan\fR. If \fIname\fR is \fBtest\fR or \fBdevel\fR, bash will be
executed with a controlling terminal and also remove the container on exit.
If \fIname\fR is \fBsvscan\fR, bash will be executed with a controlling
terminal, but the container will not be removed on exit. If \fIprog\fR is
\fBauto\fR, systemd will be loaded for debian, ubuntu, fedora, centos,
almalinux, oraclelinux, opensuse images. For alpine linux bash will be
loaded with a controlling terminal. If systemd is empty, the default entry
point for the container will be used. If \fIprog\fR is set as auto, then
\fBrunpod\fR will instruct podman or docker to execute \fBsystemd\fR(1) if
the image supports systemd or execute /sbin/init for alpine and gentoo
images.

If you need to pass any command line switches to \fBprog\fR, use \-\- to
prevent the options getting passed to \fBrunpod\fR instead of \fBprog\fR.

e.g. The below passes the timezone parameter to indimail's
\fBdocker-entrypoint\fR(8).

.EX
runpod --id=2a13c928f3aa --name=svscan \-\- -t UTC
.EE

.SH Examples with indimail, indimail-mta, indimail-web containers

indimail-mta, indimail-virtualdomains and indimail-web containers use
docker-entrypoint(8). The docker-entrypoint does few initialization and
configuration to self configure for your domain. The examples below give
you a general idea on how to use the \fBrunpod\fR command with indimail
contianers. Here --id=9f0e74d71e91 refers to an indimail-mta-web container
image.

.EX
Execute ls -l in indimail's webmail container, pass -t Asia/Kolkata to the
docker-entrypoint. Pass additional \-\- before ls -l so that the
docker-entrypoint doesn't interpret -l option.

runpod -a "--rm" --id=9f0e74d71e91 -n webmail -- \\
  -t Asia/Kolkata -- /bin/ls -l

Execute bash

runpod -a "--rm" --id=9f0e74d71e91 -n webmail -- \\
  -t Asia/Kolkata -- /bin/bash

Execute repair shell. You can fix things in the container and commit the
image before coming out

runpod -a "--rm" --id=9f0e74d71e91 -n webmail -- -r
.EE

.SH OPTIONS
.TP 2
\fB\-\-id\fR=\fIimageID\fR | -i \fIimageID\fR
Use this option to specify image id

.TP
\fB\-\-name\fR=\fIname\fR | -\fBn\fR \fIname\fR
Use this option to specify name for the running image. \fIname\fR has
special significance if name is either of \fBindimail\fR,
\fBindimail-mta\fR, \fBwebmail\fR which causes following ports to be mapped

.EX
2025 -> 25, 2106 -> 106, 2110 -> 110  2143 -> 143, 2209 -> 209,
2366 -> 366, 2465 -> 465, 2587 -> 587 2628 -> 628, 2993 -> 993,
2995 -> 995, 3110 -> 4110 3143 -> 4143, 5110 -> 9110, 5143 -> 9143,
8080 -> 80
.EE

\fIname\fR is what you see in the \fBpodman\fR ps or the \fBdocker\fR ps
command.

.TP
\fB\-\-args\fR=\fIextra_args\fR | -\fBa\fR \fIextra_args\fR
Pass extra arguments to podman or docker command. You need use this for
every extra arg that you need to supply to the podman or docker command.

.TP
\fB\-\-capability\fR=\fIcap\fR | -\fBC\fR \fIcap\fR
Use this option to add capabilities like SYS_PTRACE, SYS_ADMIN, IPC_LOCK,
SYS_RESOURCE. You can specify this option multiple times.

.TP
\fB\-\-host\fR=\fIhost\fR | -\fBh\fR \fIhost\fR
Specify hostname for the container.

.TP
\fB\-\-port\fR=\fIlport\fR:\fIrport\fR -\fBp\fR \fIlport\fR:\fIrport\fR
Map local port \fIlport\fR to remote port \fIrport\fR in the container. You
can specify this option multiple times.

.TP
\fB\-\-command\fR=\fIdocker\fR|\fIpodman\fR | -\fBc\fR \fIdocker\fR|\fIpodman\fR
Use \fBdocker\fR(1) or \fBpodman\fR(1) to run the container. Default is
\fBpodman\fR(1).

.TP
\fB\-\-volume\fR=\fIlocal_dir\fR:\fIremote_dir\fR | \fB\-v\fR
\fIlocal_dir\fR:\fIremote_dir\fR mount \fIlocal_dir\fR as \fIremote_dir\fR
in the container. You can specify this option multiple times.

.TP
\fB\-\-cgroup\fR=\fIgroup\fR | -\fBg\fR \fIgroup\fR
Specify \fIgroup\fR for cgroups. You can specify this option multiple
times.

.SH SEE ALSO
docker(1),
podman(1)
docker-entrypoint(8)


libdkim-x/NEWS
Migrated to libtool 2.x


daemontools-x/runpod.sh
#
# $Id: runpod.sh,v 1.10 2022-12-08 09:40:09+05:30 Cprogrammer Exp mbhangui $
#
usage()
{
  (
  echo "`basename $0` -i|--id imageid [-n|--name name]"
  echo "  [-a|--args extra_args1 -a|--args extra_args2 ... -a|--args extra_argsn]"
  echo "  [-h|--host host] [-g|--cgroup cgroup]"
  echo "  [-p|--port lport1:rport1 -p|--port lport2:rport2 ... -p|--port lportn:rportn]"
  echo "  [-v|--volume.dir:mount_point1 -v|--volume dir:mount_point2 ... -v|--volume dir:mount_pointn]"
  echo "  [-C|--capability cap1 -C|--capability cap2 ... -C|--capability cap3] | [-C|--capability auto]"
  echo "  [-c|--cmd podman|docker] prog args args ..."
  echo
  echo "lport - local port, rport - remote port"
  ) 1>&2
  exit $1
}

set_volume()
{
  word_count=$(echo $1 | cut --output-delimiter=" " -d: -f1,2,3 | wc -w)
  if [ $word_count -ne 2 ] ; then
    echo "volume should be of the form \"dir:mount_point\"" 1>&2
    exit 1
  fi
  dir=$(echo $1|cut -d: -f1)
  first_char=$(echo $dir | cut -c1)
  if [ " $first_char" = " /" ] ; then
    if [ ! -d $dir ] ; then
      echo "$dir: No such directory" 1>&2
      usage 1
    fi
  else
    $command volume inspect $dir > /dev/null
    if [ $? -ne 0 ] ; then
      (
      echo "$dir: No such volume"
      echo "List of volumes are"
      $command volume ls
      ) 1>&2
      usage 1
    fi
  fi
  dir=$(echo $1|cut -d: -f2)
  first_char=$(echo $dir | cut -c1)
  if [ " $first_char" != " /" ] ; then
    echo "$dir should be a valid path" 1>&2
    usage 1
  fi
  vol_args="$vol_args""-v $1 "
}

set_defaults()
{
  case $name in
    indimail|indimail-mta|webmail|devel|svscan|test|mail)
    if [ -z "$port_args" ] ; then
      case $name in
        indimail|indimail-mta|webmail)
        port_args="$port_args-p 2025:25   -p 2106:106  -p 2110:110  -p 2143:143 "
        port_args="$port_args-p 2209:209  -p 2366:366  -p 2465:465  -p 2587:587 "
        port_args="$port_args-p 2628:628  -p 2993:993  -p 2995:995  -p 3110:4110 "
        port_args="$port_args-p 3143:4143 -p 5110:9110 -p 5143:9143 -p 8080:80 -p 8081:443"

        # port_str is used just for display purpose
        NEWLINE=$'\n'
        port_str=""
        port_str="${port_str}  -p 2025:25   -p 2106:106  -p 2110:110  -p 2143:143${NEWLINE}"
        port_str="${port_str}  -p 2209:209  -p 2366:366  -p 2465:465  -p 2587:587${NEWLINE}"
        port_str="${port_str}  -p 2628:628  -p 2993:993  -p 2995:995  -p 3110:4110${NEWLINE}"
        port_str="${port_str}  -p 3143:4143 -p 5110:9110 -p 5143:9143 -p 8080:80 -p 8081:443"
      ;;
      esac
    fi
    if [ -z "$cap_args" ] ; then
      cap_args="$cap_args--cap-add SYS_PTRACE --cap-add SYS_ADMIN "
      cap_args="$cap_args--cap-add IPC_LOCK   --cap-add SYS_RESOURCE "
      cap_args="$cap_args--cap-add=NET_ADMIN  --cap-add=CAP_NET_RAW "
      cap_args="$cap_args--cap-add=SYS_NICE "
    else
      echo $cap_args|grep SYS_ADMIN > /dev/null
      if [ $? -ne 0 ] ; then
        cap_args="$cap_args--cap-add SYS_ADMIN "
      fi
    fi
    if [ -z "$host_args" ] ; then
      host_args="-h indimail.org"
    fi
    if [ -z "$vol_args" ] ; then
      if [ -d /home/$command/queue ] ; then
        vol_args="$vol_args-v /home/$command/queue:/var/queue "
      fi
      if [ -d /home/$command/mail ] ; then
        vol_args="$vol_args-v /home/$command/mail:/home/mail "
      fi
      if [ " $name" = " devel" -a  -d /usr/local/src ] ; then
        vol_args="$vol_args-v /usr/local/src:/usr/local/src"
      fi
    fi
    if [ " $name" = " devel" -o " $name" = " test" ] ; then
      if [ $# -eq 0 ] ; then
        systemd="bash"
        extra_args="$extra_args ""-ti --rm"
      fi
    elif [ " $name" = " svscan" ] ; then
        extra_args="$extra_args ""-ti"
    fi
    if [ -z "$extra_args" ] ; then
      detached=1
      extra_args="-d"
    fi
    ;;
  esac
  if [ -n "$cgroup" -a "$cgroup" = "auto" ] ; then
    cgroup="-v /sys/fs/cgroup:/sys/fs/cgroup:rw"
  fi
}

options=$(getopt -a -n runpod -o "a:n:h:c:i:p:C:g:v:" -l args:,name:,host:,cmd:,id:,port:,capa:,cgroups:,volume: -- "$@")
if [ $? != 0 ]; then
  usage
fi

command=podman
name="test"
detached=0
port_str=""
port_args=""
host_args=""
vol_args=""
cap_args=""
extra_args=""
systemd=""
imageid=""
eval set -- "$options"
while :
do
  case "$1" in
  -a | --args)
  echo $2 | grep "\-d" >/dev/null && detached=1 
  extra_args="$extra_args ""$2"
  shift 2
  ;;

  -n | --name)
  name=$2
  shift 2
  ;;

  -h | --host)
  host_args="-h $2"
  shift 2
  ;;

  -c | --cmd)
  command=$2
  shift 2
  if [ " $command" != " podman" -a " $command" != " docker" ] ; then
    echo "command should be podman or docker" 1>&2
    usage 1
  fi
  ;;

  -i | --id)
  imageid=$2
  shift 2
  ;;

  -p | --port)
  port_args="$port_args""-p $2 "
  shift 2
  ;;

  -C | --capability)
  if [ " $2" = " auto" ] ; then
    cap_args="$cap_args--cap-add SYS_PTRACE --cap-add SYS_ADMIN "
    cap_args="$cap_args--cap-add IPC_LOCK   --cap-add SYS_RESOURCE "
    cap_args="$cap_args--cap-add SYS_NICE "
  else
    cap_args="$cap_args--cap-add $2 "
  fi
  shift 2
  ;;

  -g | --cgroup)
  cgroups="$optarg"
  ;;

  -v | --volume)
  set_volume $2
  shift 2
  ;;

  --) # end of options
  shift
  break
  ;;

  *)
  echo "Unexpected option: $1 - this should'nt happen." 1>&2
  usage
  ;;
  esac
done
if [ -z "$imageid" ] ; then
  usage 1
fi
set_defaults $*
if [ $# -gt 0 -a "$1" = "auto" ] ; then
  tag=`$command images | grep  $imageid | awk '{print $2}'`
  case $tag in
    xenial*|debian*|focal*|bionic*|archlinux*|ubi*)
    systemd=/lib/systemd/systemd
    ;;
    alpine*|gentoo*)
    systemd="/sbin/init"
    ;;
    *)
    systemd=/usr/lib/systemd/systemd
    ;;
  esac
  if [ -n "$systemd" ] ; then
    set "$systemd $*"
  fi
fi
echo "$command run $cgroup $extra_args"
echo "  --publish-all --name $name"
echo "  --device /dev/fuse"
if [ -n "$host_args" ] ; then
  echo "  $host_args"
fi
if [ -n "$cap_args" ] ; then
  echo $extra_args | grep "\--privileged" >/dev/null
  if [ $? -eq 0 ] ; then
    cap_args=""
  else
    echo "  $cap_args"
  fi
fi
if [ -z "$port_args" ] ; then
  if [ -n "$port_str" ] ; then
    port_args=$port_str
  fi
fi
if [ -n "$port_args" ] ; then
  echo "  $port_args"
fi
if [ -n "$vol_args" ] ; then
  echo "  $vol_args"
fi
echo "  $imageid $*"

$command run $cgroup $extra_args \
  --publish-all --name $name \
  --device /dev/fuse \
  $host_args $cap_args $port_args $vol_args \
  $imageid $*
if [ $? -eq 0 ] ; then
  if [ $detached -eq 1 ] ; then
    $command exec -ti $name bash
  fi
else
  echo "Removing container id $name" 1>&2
  $command rm $name > /dev/null 2>&1
fi

#
# $Log: runpod.sh,v $
# Revision 1.10  2022-12-08 09:40:09+05:30  Cprogrammer
# fixed setting of systemd
#
# Revision 1.9  2022-12-07 20:51:56+05:30  Cprogrammer
# added port 8081 for mapping to port 443 on container OS
#
# Revision 1.8  2022-12-07 00:56:36+05:30  Cprogrammer
# run in foreground if name is test or devel
#
# Revision 1.7  2022-11-10 16:55:40+05:30  Cprogrammer
# redirect all error message to fd 2
#
# Revision 1.6  2022-11-10 12:49:11+05:30  Cprogrammer
# fixed extra_args variable getting clobbered
#
# Revision 1.5  2022-11-08 23:27:52+05:30  Cprogrammer
# added case for name=svscan
#
# Revision 1.4  2022-11-07 22:43:38+05:30  Cprogrammer
# prevent output of test on stdout
#
# Revision 1.3  2022-11-01 23:53:11+05:30  Cprogrammer
# fixed port clash
#
# Revision 1.2  2022-09-19 21:21:22+05:30  Cprogrammer
# added SYS_NICE capability to defaults
#
# Revision 1.1  2022-09-12 22:50:12+05:30  Cprogrammer
# Initial revision
#
#
